### ISSUE TITLE
Web browsers based on webkit2gtk are not able to stream youtube videos properly

### ISSUE THREAD

**Issue by egorenar** [void-linux/void-packages#25852]
### System

* xuname:  
  Void 5.8.16_1 x86_64 GenuineIntel notuptodate rrrrrmmdnFFFFFF
* package:  
  * webkit2gtk-2.30.0_1
  * gstreamer1-1.16.2_1

### Expected behavior

Playing of youtube videos with webkit2gtk works.

### Actual behavior

* Playback hangs after fast forwarding.
* Pausing/Unpausing a video hangs.
* Playing multiple youtube videos simultaneously is broken.

### Steps to reproduce the behavior

1. Execute `/usr/libexec/webkit2gtk-4.0/MiniBrowser`
2. Go to www.youtube.com
3. Choose any video and start playing
4. Fast forward to an arbitrary time point

### Notes

* The same behavior can also be observed with **surf** or **epiphany**.
* **It might be a gstreamer issue (which is used for youtube streaming) and not a webkit2gtk one.**
* Updating webkit2gtk to version 2.30.2 didn't help.

**Comment by minikN** [comment]
Thanks @egorenar .

This seems to be related: https://github.com/atlas-engineer/nyxt/issues/1017

**Comment by travankor** [comment]
https://github.com/void-linux/void-packages/pull/25313 might fix this

cc @not-chicken

**Comment by not-chicken** [comment]
@travankor thanks for the ping. 
Unfortunately, I can still reproduce this with GStreamer 1.18 :(

**Comment by minikN** [comment]
I just updated to webkit2gtk 2.30.2_1 and this issue still persists.

```
003 ~ ➜ xbps-query -Rs gstreamer1-
[*] gstreamer1-1.16.2_1                Core GStreamer libraries and elements (1.x)
[*] gstreamer1-devel-1.16.2_1          Core GStreamer libraries and elements (1.x) - development files
[-] gstreamer1-pipewire-0.3.12_1       Server and user space API to deal with multimedia pipelines - gstreamer plugin
[*] gstreamer1-32bit-1.16.2_1          Core GStreamer libraries and elements (1.x) (32bit)
[-] gstreamer1-devel-32bit-1.16.2_1    Core GStreamer libraries and elements (1.x) - development files (32bit)
[-] gstreamer1-pipewire-32bit-0.3.12_1 Server and user space API to deal with multimedia pipelines - gstreamer plugin (32bit)
```

```
239 ~ ➜ xbps-query -Rs webkit2gtk
[*] webkit2gtk-2.30.2_1             GTK+3 port of the WebKit2 browser engine
[*] webkit2gtk-devel-2.30.2_1       GTK+3 port of the WebKit2 browser engine - development files
[-] webkit2gtk-32bit-2.30.2_1       GTK+3 port of the WebKit2 browser engine (32bit)
[-] webkit2gtk-devel-32bit-2.30.2_1 GTK+3 port of the WebKit2 browser engine - development files (32bit)
```

**Comment by notthewave** [comment]
Interesting that the epiphany flatpak does not have this problem...

**Comment by ericonr** [comment]
Is this still an issue? We have removed sndio from gstreamer.

**Comment by minikN** [comment]
> Is this still an issue? We have removed sndio from gstreamer.

This issue still exists for me.

**Comment by ericonr** [comment]
I couldn't even get into youtube properly. I will try to update my gstreamer installation and see what happens.

**Comment by egorenar** [comment]
MiniBrowser still having issues with Youtube, sometimes i can't even open youtube :(

**Comment by travankor** [comment]
try the #27996 branch if you can

**Comment by ericonr** [comment]
@travankor still behaving weirdly; audio seems better, but video is extremely locked up (at points more than 10 seconds without moving).

**Comment by travankor** [comment]
The reason why video is choppy is most likely due to `gstreamer-vaapi`.
I can't get it work correctly, so I have uninstalled it. Pure software video decoding seems to work better with webkit2gtk.

**Comment by travankor** [comment]
@egorenar Please try the above suggestions and report back. I can't reproduce this anymore after updating gstreamer and removing `gstreamer-vaapi`

```
$ xbps-query -s gst
[*] gst-libav-1.18.3_1             GStreamer libav plugin (1.x)
[*] gst-plugins-bad1-1.18.3_1      GStreamer plugins from the bad set (v1.x)
[*] gst-plugins-base1-1.18.3_1     GStreamer Base Plug-ins (v1.x)
[*] gst-plugins-good1-1.18.3_1     GStreamer set of well-maintained good plu...
[*] gst-plugins-ugly1-1.18.3_1     GStreamer plugins from the ugly set (1.x)
[*] gst1-editing-services-1.18.3_1 GStreamer 1.x Editing Services
[*] gstreamer1-1.18.3_1            Core GStreamer libraries and elements (1.x)
[*] gstreamer1-pipewire-0.3.19_1   Server and user space API to deal with mu...
```

**Comment by egorenar** [comment]
I tried out @travankor 's  branch but still having the same issues, youtube videos just hang when i try fast forwarding.
Tested with MiniBrowser and Next.

```
➜  void-packages git:(gstreamer) xbps-query -s gst
[*] gst-libav-1.18.3_1               GStreamer libav plugin (1.x)
[*] gst-plugins-bad1-1.18.3_1        GStreamer plugins from the bad set (v1.x)
[*] gst-plugins-base1-1.18.3_1       GStreamer Base Plug-ins (v1.x)
[*] gst-plugins-base1-32bit-1.16.2_2 GStreamer Base Plug-ins (v1.x) (32bit)
[*] gst-plugins-good1-1.18.3_1       GStreamer set of well-maintained good plug-ins (1.x)
[*] gst-plugins-ugly1-1.18.3_1       GStreamer plugins from the ugly set (1.x)
[*] gst1-editing-services-1.18.3_1   GStreamer 1.x Editing Services
[*] gstreamer1-1.18.3_1              Core GStreamer libraries and elements (1.x)
[*] gstreamer1-32bit-1.16.2_2        Core GStreamer libraries and elements (1.x) (32bit)
[*] gstreamer1-pipewire-0.3.20_1     Server and user space API to deal with multimedia pipelines - gstreamer plugin
[*] gstreamermm-1.10.0_2             GStreamer API C++ bindings
```

**Comment by kawaiiamber** [comment]
The same issue can currently be reproduced with the `surf` package as well. It can play a video... kind of; it will freeze whenever one tries to pause / play, and sometimes it will just freeze altogether without any action. Video streaming is very inconsistant with `webkit2gtk` package.

**Comment by notthewave** [comment]
Just checked my alpine install. This problem only seems to occur on void. Skipping in a video after installing gst-plugins-* seems to work fine on alpine.

**Comment by kawaiiamber** [comment]
I now think it is most likely an issue with the `gstreamer1` package. I notice on apps that use the `qt5-multimedia` package (which uses `gstreamer1` package) suck as `nheko`, video playback is also very inconsistent in those cases as well. Video fails to play in the `gtk4-demo` package, which uses `gstreamer1`.

**Comment by kawaiiamber** [comment]
I tried messing around with different package managers and seeing how they package `gstreamer`, and thought I had something, but it didn't seem to change anything. I'm really scratching my head why the `gstreamer1` package is bugged...

**Comment by ericonr** [comment]
Is it actually a bug on our side or just webkit2gtk in general?

**Comment by kawaiiamber** [comment]
I noticed that trying to run `gst-play-1.0 <video-file>` on either local video files or links to video files online reproduces the same inconsistant result that crashes gstreamer. I can't seem to get a good error as to why it's freezing whenever one fast forwards / rewinds. Doing `gst-play-1.0 --verbose <file/url>` also doesn't give information about why it's freezing - all of the output to the console also freezes.
In short - this is definitely an issue with the `gstreamer1` package as the `gst-play-1.0` testing tool crashes. To reproduce the bug: `gst-play-1.0 <file/url>`, then fast forward or rewind. I'm not too familiar with Linux stuff, and thus don't know a better way to see why it's freezing - e.g. memory leak, etc.

**Comment by kawaiiamber** [comment]
I decided to try and follow some steps from [here](https://askubuntu.com/questions/15437/why-is-my-terminal-freezing-up). I reproduced the bug with `gst-play-1.0 <file>` and tracked forward. Then, ran `ps axo pid,wchan:32,cmd`. The line that relates to the `gst-play-1.0` command crash:
```
12199 -                                gst-play-1.0 /home/amber/vids/blackbear - queen of broken hearts.mkv
```
On the column for `wchan`, it was an empty `-`. So next, I used the `strace` package/command on that PID to try and trace it with the following result:
```
[amber@otaku ~]$ doas strace -p 12199              
doas (amber@otaku) password: 
strace: Process 12199 attached
futex(0x7ffe086147c4, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x56022f252164, FUTEX_LOCK_PI_PRIVATE, NULL
```
and the command is stuck there. I noticed [this](https://stackoverflow.com/questions/39920801/resource-temporarily-unavailable-futex-while-straceing-sqoop-command), which seemed somewhat related. The output of `doas strace -f -p 12199` is [here](https://pastebin.com/YU7YEuLx) (I stopped the command with ^C signal as it was repeating over and over). These two lines seemed interesting:
```
[pid 12203] nanosleep({tv_sec=0, tv_nsec=50000000}, 0x7f713fb56880) = 0
[pid 12203] recvmsg(9, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
```
I'm not really sure if / how the pid of 12203 relates (nothing returned with `ps aux | grep -v grep | grep 12203`). That's all I'm able t get so far...

**Comment by kawaiiamber** [comment]
I've been messing around trying to build `gstreamer1` myself and changing the template locally for certain packages. I think I have something I want to test. Possibly dumb question, but, is there a way for me to have my system use my built of `gstreamer1` that was biult from the template I wrote instead of the one from the standard repo? I can't just `xbps-remove gstreamer1` to then `xbps-install --repository=./hostdir gstreamer1` as it won't let me initially remove `gstreamer1`.

**Comment by ericonr** [comment]
Install `xtools` and do `xi -f gstreamer1 gst-libav`

**Comment by kawaiiamber** [comment]
> Install `xtools` and do `xi -f gstreamer1 gst-libav` (after building the packages with changes, of course).

Thank you ~

**Comment by st3r4g** [comment]
> I noticed that trying to run gst-play-1.0 <video-file> on either local video files or links to video files online reproduces the same result that crashes gstreamer.

Does it happen with this [test video](https://www.freedesktop.org/software/gstreamer-sdk/data/media/sintel_trailer-480p.webm)? Playing this with `gst-play-1.0` works fine for me...

**Comment by travankor** [comment]
>Does it happen with this test video? Playing this with gst-play-1.0 works fine for me...

Same `gst-play-1.0` works but a webkit2gtk browser freezes for this file.

**Comment by st3r4g** [comment]
`/usr/libexec/webkit2gtk-4.0/MiniBrowser` from the `webkit2gtk` package also plays the test video fine for me. I suspect the issue arises when installing some plugin package

**Comment by kawaiiamber** [comment]
> `/usr/libexec/webkit2gtk-4.0/MiniBrowser` from the `webkit2gtk` package also plays the test video fine for me. Maybe the issue arises when installing some plugin package? If I install `gst-libav` (required to play youtube videos) and try to seek/pause a youtube video, it hangs. But the test video still works.

I guess I should have been more specific. The video will play with `gst-play-1.0`, but will freeze upon doing any action (a list of them are available with hitting `k`).

**Comment by st3r4g** [comment]
> The video will play with gst-play-1.0, but will freeze upon doing any action (a list of them are available with hitting k).

they work for me

**Comment by kawaiiamber** [comment]
> > The video will play with gst-play-1.0, but will freeze upon doing any action (a list of them are available with hitting k).
> 
> they work for me

Interesting - they don't for me. I'm using musl and have the `gst-plugins-good1` and `gst-plugins-bad1` installed, as well. Is it only the demo video that works for you? Or any video?

**Comment by st3r4g** [comment]
Then it might be a different issue, maybe musl related.

> Or any video?

Didn't test many, but any video works.

**Comment by matheson-581** [comment]
This issue is still ongoing. Every webkit2gtk-based browser I have tested (badwolf, netsurf, surf, epiphany) fails in the exact same way: skipping an ad, pausing, rewinding or fast-forwarding on YouTube, Odysee, DTube, PeerTube or Dailymotion causes the video stream to hang and the entire browser tab to freeze. 

Interestingly, playing, pausing, rewinding and fast-forwarding VP8 webm files and Ogg Theora ogv files from Wikimedia works correctly without issue - the problem only seems to affect h264/h265 streams. 

Playing, pausing, rewinding, and fast-forwarding the above-referenced VP8 test file with gst-play-1.0 locally works perfectly.  Playing, pausing, rewinding, and fast-forwarding a h264 test file with gst-play-1.0 locally works perfectly. Playing, pausing, rewinding, and fast-forwarding a direct video URI from DTube (right click the video, click "copy video link location") with gst-play-1.0 also works perfectly. Playing, pausing, rewinding, and fast-forwarding the same direct video URI in the browser also works perfectly. However, when playing the video from the DTube website (i.e., just loading the DTube URL and clicking play instead of right clicking and getting the video link location), the video hangs and the tab freezes when pausing, fast-forwarding or rewinding. Quite frustrating. This does not affect non-webkit browsers such as Firefox, Chromium, Basilisk, or Pale Moon; those browsers obviously do not use gstreamer for their backend, which seems to be the source of the problem.

My installation: kernel 5.11.18_1, gstreamer1-1.18.4_1, gstreamer-vaapi-1.18.4_1, gst-libav-1.18.4_1, gst-omx-1.18.4_1, gst-plugins-base1-1.18.4_1, gst-plugins-good1-1.18.4_1, gst-plugins-bad1-1.18.4_3, gst-plugins-ugly1-1.18.4_2, webkit2gtk-2.32.0_3

**Comment by egorenar** [comment]
I built nyxt 2.0.0 today and youtube works great :+1: finally.

**Comment by kawaiiamber** [comment]
Not sure what changed recently, but video playback seems to work ok now.

**Comment by egorenar** [comment]
```
$ xbps-query -s gst
[*] gst-libav-1.18.4_1               GStreamer libav plugin (1.x)
[*] gst-plugins-bad1-1.18.4_3        GStreamer plugins from the bad set (v1.x)
[*] gst-plugins-base1-1.18.4_1       GStreamer Base Plug-ins (v1.x)
[*] gst-plugins-base1-32bit-1.18.4_1 GStreamer Base Plug-ins (v1.x) (32bit)
[*] gst-plugins-good1-1.18.4_1       GStreamer set of well-maintained good plug-ins (1.x)
[*] gst-plugins-ugly1-1.18.4_2       GStreamer plugins from the ugly set (1.x)
[*] gst1-editing-services-1.18.4_1   GStreamer 1.x Editing Services
[*] gstreamer1-1.18.4_2              Core GStreamer libraries and elements (1.x)
[*] gstreamer1-32bit-1.18.4_2        Core GStreamer libraries and elements (1.x) (32bit)
[*] gstreamer1-pipewire-0.3.28_1     Server and user space API to deal with multimedia pipelines - gstreamer plugin
[*] gstreamermm-1.10.0_2             GStreamer API C++ bindings
```

**Comment by kawaiiamber** [comment]
> not for me...
> 
> ```
> ~ ❱ xbps-query -s gst
> [*] gst-plugins-bad1-1.18.4_3    GStreamer plugins from the bad set (v1.x)
> [*] gst-plugins-base1-1.18.4_1   GStreamer Base Plug-ins (v1.x)
> [*] gst-plugins-good1-1.18.4_1   GStreamer set of well-maintained good plug-ins (1.x)
> [*] gst-plugins-ugly1-1.18.4_2   GStreamer plugins from the ugly set (1.x)
> [*] gstreamer1-1.18.4_2          Core GStreamer libraries and elements (1.x)
> [*] gstreamer1-pipewire-0.3.28_1 Server and user space API to deal with multimedia pipelines - gs...
> ```

You need `gst-libav`

**Comment by notthewave** [comment]
still not working. I'm on musl.

**Comment by egorenar** [comment]
It works for me also if i remove `gst-libav` package

**Comment by ericonr** [comment]
This seems to be solved on my side as well.

**Comment by travankor** [comment]
I think this was fixed with webkit2gtk>=2.32.1.

Although I see `ALSA lib pcm_dmix.c:1075:(snd_pcm_dmix_open) unable to open slave` when using webkit2gtk apps with another app that tries to use audio at the same time.

**Comment by ericonr** [comment]
I suppose it's using ALSA first before falling back to pulse? It's another issue, if it is one at all. Could be gstreamer to blame too.

**Comment by ericonr** [comment]
Thanks everyone for the investigative help!

**Comment by q66** [comment]
this is not really fixed

**Comment by notthewave** [comment]
still does not work for me

**Comment by q66** [comment]
seems to be this issue https://github.com/void-linux/void-packages/issues/15631

disabling the pulse plugin in gst-plugins-good fixes it, so this is a problem with either the gst pulse plugin or libpulse itself

**Comment by q66** [comment]
Fixed properly in https://github.com/void-linux/void-packages/commit/53d00fffc2423f0647c8b5f58d1fd25c55ff565f
